File: src/database/conexao.php
<?php
require_once 'config.php';

class Database
{
    private static $pdo = null;

    public static function getConnection()
    {
        if (self::$pdo === null) {
            try {
                // Primeiro conecta sem especificar o banco de dados para criá-lo se necessário
                $dsn = "mysql:host=" . DB_HOST . ";charset=utf8mb4";
                $options = [
                    PDO::ATTR_ERRMODE => PDO::ERRMODE_EXCEPTION,
                    PDO::ATTR_DEFAULT_FETCH_MODE => PDO::FETCH_ASSOC,
                    PDO::ATTR_EMULATE_PREPARES => false,
                ];

                self::$pdo = new PDO($dsn, DB_USER, DB_PASS, $options);

                // Verifica e cria o banco de dados se não existir
                self::createDatabase();

                // Agora conecta ao banco de dados específico
                $dsn = "mysql:host=" . DB_HOST . ";dbname=" . DB_NAME . ";charset=utf8mb4";
                self::$pdo = new PDO($dsn, DB_USER, DB_PASS, $options);

                // Cria as tabelas se não existirem
                self::createTables();
            } catch (PDOException $e) {
                die("Erro na conexão com o banco de dados: " . $e->getMessage());
            }
        }
        return self::$pdo;
    }

    private static function createDatabase()
    {
        try {
            // Verifica se o banco de dados já existe
            $stmt = self::$pdo->query("SELECT SCHEMA_NAME FROM INFORMATION_SCHEMA.SCHEMATA WHERE SCHEMA_NAME = '" . DB_NAME . "'");

            if ($stmt->rowCount() == 0) {
                // Cria o banco de dados
                self::$pdo->exec("CREATE DATABASE " . DB_NAME . " CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci");
                echo "Banco de dados criado com sucesso!<br>";
            }

            // Seleciona o banco de dados
            self::$pdo->exec("USE " . DB_NAME);
        } catch (PDOException $e) {
            die("Erro ao criar o banco de dados: " . $e->getMessage());
        }
    }

    private static function createTables()
    {
        global $database_structure;

        try {
            foreach ($database_structure as $table => $structure) {
                // Verifica se a tabela já existe
                $stmt = self::$pdo->query("SHOW TABLES LIKE '$table'");

                if ($stmt->rowCount() == 0) {
                    // Monta a query de criação da tabela
                    $columns = [];
                    foreach ($structure['columns'] as $name => $definition) {
                        $columns[] = "$name $definition";
                    }

                    $query = "CREATE TABLE $table (" . implode(', ', $columns);

                    // Adiciona chaves estrangeiras se existirem
                    if (isset($structure['foreign_keys'])) {
                        foreach ($structure['foreign_keys'] as $column => $fk) {
                            $query .= ", FOREIGN KEY ($column) REFERENCES {$fk['references']}";
                            if (isset($fk['on_delete'])) {
                                $query .= " ON DELETE {$fk['on_delete']}";
                            }
                        }
                    }

                    $query .= ") {$structure['options']}";

                    // Executa a criação da tabela
                    self::$pdo->exec($query);
                    echo "Tabela $table criada com sucesso!<br>";
                }
            }
        } catch (PDOException $e) {
            die("Erro ao criar tabelas: " . $e->getMessage());
        }
    }

    // Método para fechar a conexão (útil para testes)
    public static function closeConnection()
    {
        self::$pdo = null;
    }
}

// Obtém a conexão PDO
$pdo = Database::getConnection();

File: src/database/conexao.py
import mysql.connector
from mysql.connector import Error
from .config import DB_CONFIG


def get_db_connection():
    try:
        connection = mysql.connector.connect(**DB_CONFIG)
        return connection
    except Error as e:
        print(f"Erro ao conectar ao MySQL: {e}")
        return None


def create_database():
    try:
        # Conectar sem especificar o banco de dados
        conn = mysql.connector.connect(
            host=DB_CONFIG["host"],
            user=DB_CONFIG["user"],
            password=DB_CONFIG["password"],
        )
        cursor = conn.cursor()

        # Criar banco de dados se não existir
        cursor.execute(
            """
        CREATE DATABASE IF NOT EXISTS reconhecimento_facial 
        CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci
        """
        )

        # Usar o banco de dados
        cursor.execute("USE reconhecimento_facial")

        # Criar tabela de cadastros
        cursor.execute(
            """
        CREATE TABLE IF NOT EXISTS cadastros (
            id INT AUTO_INCREMENT PRIMARY KEY,
            nome VARCHAR(100) NOT NULL,
            sobrenome VARCHAR(100) NOT NULL,
            apelido VARCHAR(100),
            observacoes TEXT,
            data_cadastro DATETIME NOT NULL
        ) ENGINE = InnoDB DEFAULT CHARSET = utf8mb4 COLLATE = utf8mb4_unicode_ci
        """
        )

        # Criar tabela de imagens
        cursor.execute(
            """
        CREATE TABLE IF NOT EXISTS imagens_cadastro (
            id INT AUTO_INCREMENT PRIMARY KEY,
            cadastro_id INT NOT NULL,
            caminho_imagem VARCHAR(255) NOT NULL,
            data_upload DATETIME NOT NULL,
            CONSTRAINT fk_cadastro_id FOREIGN KEY (cadastro_id) 
            REFERENCES cadastros(id) ON DELETE CASCADE
        ) ENGINE = InnoDB DEFAULT CHARSET = utf8mb4 COLLATE = utf8mb4_unicode_ci
        """
        )

        conn.commit()
        print("Banco de dados e tabelas criados com sucesso.")

    except Error as e:
        print(f"Erro ao criar banco de dados: {e}")
    finally:
        if conn and conn.is_connected():
            cursor.close()
            conn.close()
